File: Java/1525.NumberOfGoodWaysToSplitAString.java
class Solution {
  public int numSplits(String s) {
    // storing distinct characters frequency of left & right half
    int[] leftHalf = new int[26];
    int[] rightHalf = new int[26];

    for (char c : s.toCharArray()) {
      ++rightHalf[c - 'a'];
    }

    int count = 0;
    for (int i = 0; i < s.length(); ++i) {
      char ch = s.charAt(i);
      ++leftHalf[ch - 'a'];
      --rightHalf[ch - 'a'];

      // checking if it is a good split
      if (isGoodSplit(leftHalf, rightHalf))
        ++count;
    }

    return count;
  }

  /*
   * A good split is a split where the left & the right halves have same number of distinct characters.
   */
  private static boolean isGoodSplit(int[] lh, int[] rh) {
    int lDistinct = 0, rDistinct = 0;
    for (int i = 0; i < 26; ++i) {
      lDistinct += lh[i] > 0 ? 1 : 0;
      rDistinct += rh[i] > 0 ? 1 : 0;
    }

    return lDistinct == rDistinct;
  }
}

File: C++/1525. NumberOfGoodWaysToSplitAString.cpp
/*
You are given a string s.
A split is called good if you can split s into 2 non-empty strings p and q where its concatenation is equal to s and the number of distinct letters in p and q are the same.

Return the number of good splits you can make in s.
*/

class Solution {
public:
    int numSplits(string s) {
        int n = s.size();
        int ans=0;
        // Map for having freq of each char in set2
        unordered_map<char, int> freq;
        
        // Total distinct elements in string - keeps count for split 2
        unordered_set<char> s2;
        for(int i=0;i<n;i++){
            s2.insert(s[i]);
            freq[s[i]]++;
        }
        int distElt = s2.size();
        
        int dist2=0;
        // Set for having count of distinct elements in 1st split - keeps running count
        unordered_set<char> s1;
        for(int i=0;i<n;i++){
            s1.insert(s[i]);
            dist2 = s1.size();
            freq[s[i]]--;
            if(freq[s[i]]==0) distElt--;
            if(dist2 == distElt) ans++;
        }
        return ans;
    }
};